A Draft PR so Jackson-senpai can review

litehed opened the pull request:
Before issuing a pull request, please see the contributing page.

litehed committed:
Added FtcDashboard to ControllerActivity

litehed committed:
Shhhh no one will ever know

litehed committed:
Fixed many of the tests and finished tuning robot for now

my life be like oooh ahhhhhh

litehed committed:
Added Intake and Shooter subsystems and commands

Added commands and subsystems for both the intake. Next up is the flicker and the wobble arm and grabber.

JIceberg commented:
❤️

litehed committed:
Added Flicker

I hope I did it right,, I kind of just rushed it at the end

litehed committed:
Changed Run command to and Instant

JIceberg commented on TeamCode/src/main/java/org/firstinspires/ftc/teamcode/TeleMain.java:
```diff
@@ -47,16 +71,40 @@ public void initialize() {
 
+        //FlickerAction
+        flickerAction = new TimedAction(
+                ()-> flicker.setPosition(0.27),
+                ()-> flicker.setPosition(0.5),
```
Wrong way, these should be flipped. The `onEnd` should be the resting position.

JIceberg commented on TeamCode/src/main/java/org/firstinspires/ftc/teamcode/TeleMain.java:
```diff
@@ -47,16 +71,40 @@ public void initialize() {
         m_driverOp.getGamepadButton(GamepadKeys.Button.Y)
                 .toggleWhenPressed(()->mult = 0.5, ()->mult = 1.0);
 
+        m_driverOp.getGamepadButton(GamepadKeys.Button.A).toggleWhenActive(shootCommandGroup);
```
Why is this toggle when active? It should be a `whileHeld`.

JIceberg replied:
The shooter command doesn't end because it doesn't have an `isFinished()`, so this sequential command group will never end.

JIceberg commented on TeamCode/src/main/java/org/firstinspires/ftc/teamcode/TeleMain.java:
```diff
@@ -47,16 +71,40 @@ public void initialize() {
 
+        m_driverOp.getGamepadButton(GamepadKeys.Button.A).toggleWhenActive(shootCommandGroup);
+
+        m_driverOp.getGamepadButton(GamepadKeys.Button.RIGHT_BUMPER).toggleWhenActive(intakeCommand);
+        m_driverOp.getGamepadButton(GamepadKeys.Button.LEFT_BUMPER).toggleWhenActive(outtakeCommand);
```
The should be `toggleWhenPressed`, not `toggleWhenActive`. No need for the trigger binding.

JIceberg commented on TeamCode/src/main/java/org/firstinspires/ftc/teamcode/commands/Com_Shooter.java:
```diff
@@ -0,0 +1,24 @@
+    }
+
+    public void execute(){
+        shooterSubsystem.flickReset();
```
Move this to an `initialize()` and add an `isFinished()` that returns `!shooterSubsystem.flickerRunning()` which returns `flicker.running()`. If you don't move this line to `initialize()`, it will constantly reset and never end (and also never work).

JIceberg commented on TeamCode/src/main/java/org/firstinspires/ftc/teamcode/TeleMain.java:
```diff
@@ -47,16 +71,40 @@ public void initialize() {
         m_driverOp.getGamepadButton(GamepadKeys.Button.Y)
                 .toggleWhenPressed(()->mult = 0.5, ()->mult = 1.0);
 
+        m_driverOp.getGamepadButton(GamepadKeys.Button.A).toggleWhenActive(shootCommandGroup);
```
You need to make your own SequentialCommandGroup subclass that overrides `end(interrupted)` and stops the shooter when it gets let go. And then make the binding `whenHeld`.

JIceberg commented on TeamCode/src/main/java/org/firstinspires/ftc/teamcode/subsystems/ShooterSubsystem.java:
```diff
@@ -0,0 +1,51 @@
+    }
+
+    public void flickReset(){
+        if (timedAction.running())
```
```suggestion
        if (!timedAction.running())
```

JIceberg committed:
Fix typo in shooter subsystem

litehed committed:
Fixed shooter/flicker command

There is something I wanted to test, can I use a method on line 31 of my command in my SequentialGroup

JIceberg commented on TeamCode/src/main/java/org/firstinspires/ftc/teamcode/commands/groups/SequentialShooter.java:
```diff
@@ -0,0 +1,28 @@
+
+public class SequentialShooter extends SequentialCommandGroup {
+
+    private InstantCommand runFlyWheelCommand;
+    private WaitCommand waitCommand;
+    private Com_Shooter shooterCommand;
```
These are unneeded. Simply add the commands directly in the constructor.

JIceberg commented on TeamCode/src/main/java/org/firstinspires/ftc/teamcode/commands/groups/SequentialShooter.java:
```diff
@@ -0,0 +1,28 @@
+    private WaitCommand waitCommand;
+    private Com_Shooter shooterCommand;
+
+    public SequentialShooter(InstantCommand commandOne, WaitCommand commandTwo,
+                             Com_Shooter commandThree) {
+        runFlyWheelCommand = commandOne;
+        waitCommand = commandTwo;
+        shooterCommand = commandThree;
+
+        addCommands(runFlyWheelCommand, waitCommand, shooterCommand);
+    }
```
```suggestion
    public SequentialShooter(InstantCommand runFlyWheelCommand, WaitCommand waitCommand,
                             Com_Shooter shooterCommand) {
        addCommands(runFlyWheelCommand, waitCommand, shooterCommand);
    }
```

JIceberg commented on TeamCode/src/main/java/org/firstinspires/ftc/teamcode/commands/Com_Shooter.java:
```diff
@@ -0,0 +1,34 @@
+        shooterSubsystem.flick();
+    }
+
+    @Override
+    public boolean isFinished() {
+        return true;
+    }
```
Remove this.

JIceberg commented on TeamCode/src/main/java/org/firstinspires/ftc/teamcode/commands/groups/SequentialShooter.java:
```diff
@@ -0,0 +1,28 @@
+    }
+
+    @Override
+    public void end(boolean interrupted){
+        shooterCommand.stopShooter();
+    }
```
Set the flicker to the resting position here as well.

litehed committed:
Made even more fixes

hi :)

JIceberg commented on TeamCode/src/main/java/org/firstinspires/ftc/teamcode/commands/groups/SequentialShooter.java:
```diff
@@ -8,21 +8,18 @@
     @Override
     public void end(boolean interrupted){
         shooterCommand.stopShooter();
+        shooterCommand.resetFlicker();
```
that doesn't return it to the resting position, it just resets the timer and sets it to `RUNNING` mode.

JIceberg replied:
You need to actually set the position of the servo to the resting the position directly.

litehed committed:
This is why I need to actually read what the class does

JIceberg commented on TeamCode/src/main/java/org/firstinspires/ftc/teamcode/TeleMain.java:
```diff
@@ -47,18 +72,41 @@ public void initialize() {
 
+        m_driverOp.getGamepadButton(GamepadKeys.Button.A).whenHeld(shootCommandGroup);
+
+        m_driverOp.getGamepadButton(GamepadKeys.Button.RIGHT_BUMPER).whenHeld(intakeCommand);
+        m_driverOp.getGamepadButton(GamepadKeys.Button.LEFT_BUMPER).whenHeld(outtakeCommand);
```
Change these back to toggleWhenPressed

litehed replied:
no

JIceberg replied:
Understandable.

PrintingAddict committed:
heyo ethan
